drivers/sensor/sgp40/sgp40.c

```diff
@@ -0,0 +1,310 @@
+#include "sgp40.h"
+
+LOG_MODULE_REGISTER(SGP40, CONFIG_SENSOR_LOG_LEVEL);
+
+static uint8_t sgp40_compute_crc(uint16_t value)
+{
+	uint8_t buf[2];
+	sys_put_be16(value, buf);
+
+	return crc8(buf, 2, SGP40_CRC_POLY, SGP40_CRC_INIT, false);
+}
+
+static int sgp40_write_command(const struct device *dev, uint16_t cmd)
+{
+
+	const struct sgp40_config *cfg = dev->config;
+	uint8_t tx_buf[2];
+	sys_put_be16(cmd, tx_buf);
+
+	return i2c_write(cfg->bus, tx_buf, sizeof(tx_buf),
+			 cfg->i2c_addr);
+}
+
+static int sgp40_start_measurement(const struct device *dev)
+{
+	const struct sgp40_config *cfg = dev->config;
+	struct sgp40_data *data = dev->data;
+
+	uint8_t tx_buf[8] = { SGP40_CMD_MEASURE_RAW >> 8 ,
+		SGP40_CMD_MEASURE_RAW & 0xFF,
+		data->rh_param[0], data->rh_param[1], data->rh_param[2],
+		data->t_param[0], data->t_param[1], data->t_param[2],
+	};
+
+	return i2c_write(cfg->bus, tx_buf, sizeof(tx_buf),
+			 cfg->i2c_addr);
+}
+
+/*
+ * @brief Set the params for RH/T compensation
+ */
+static int sgp40_update_compensation_params(const struct device *dev,
+			   int16_t t, int16_t rh)
+{
+	struct sgp40_data *data = dev->data;
+	int16_t rh_ticks, t_ticks;
+
+	rh = CLAMP(rh, SGP40_COMP_MIN_RH, SGP40_COMP_MAX_RH);
+	t = CLAMP(t, SGP40_COMP_MIN_T, SGP40_COMP_MAX_T);
+
+	/*
+	 * Temperature and RH conversion to ticks as explained in datasheet
+	 * in section "I2C commands"
+	 */
+	rh_ticks = (rh * 0xFFFF / 100U) + 0.5;
+	t_ticks = ((t + 45U) * 0xFFFF / 175U) + 0.5;
+
+	sys_put_be16(rh_ticks, data->rh_param);
+	data->rh_param[2] = sgp40_compute_crc(rh_ticks);
+
+	sys_put_be16(t_ticks, data->t_param);
+	data->t_param[2] = sgp40_compute_crc(t_ticks);
+
+	return 0;
+}
+
+int sgp40_update_compensation_data(const struct device *dev,
+		struct sgp40_compensation_data comp_data)
+{
+	return sgp40_update_compensation_params(dev, comp_data.t, comp_data.rh);
+}
+
+int sgp40_disable_compensation(const struct device *dev)
+{
+	return sgp40_update_compensation_params(dev,
+			SGP40_COMP_DEFAULT_T, SGP40_COMP_DEFAULT_RH);
+}
+
+static int sgp40_selftest(const struct device *dev)
+{
+	const struct sgp40_config *cfg = dev->config;
+	uint8_t rx_buf[3];
+	uint16_t raw_sample;
+	int rc = 0;
+
+	rc = sgp40_write_command(dev, SGP40_CMD_MEASURE_TEST);
+	if (rc) {
+		LOG_ERR("Failed to start selftest!");
+		return rc;
+	}
+
+	k_sleep(K_MSEC(SGP40_TEST_WAIT_MS));
+
+	rc = i2c_read(cfg->bus, rx_buf, sizeof(rx_buf), cfg->i2c_addr);
+	if (rc) {
+		LOG_ERR("Failed to read data sample.");
+		return rc;
+	}
+
+	raw_sample = sys_get_be16(rx_buf);
+	if (sgp40_compute_crc(raw_sample) != rx_buf[2]) {
+		LOG_ERR("Received invalid CRC from selftest.");
+		return -EIO;
+	}
+
+	if (raw_sample != SGP40_TEST_OK) {
+		LOG_ERR("Selftest failed.");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int sgp40_attr_set(const struct device *dev, enum sensor_channel chan,
+			   enum sensor_attribute attr,
+			   const struct sensor_value *val)
+{
+	return -ENOTSUP;
+}
```

Comment:
This is not needed, the API allows attr_set to be NULL

Reply:
removed the relevant code